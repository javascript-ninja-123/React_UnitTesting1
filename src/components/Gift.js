import React,{Component} from 'react';
import { Input ,Form, Icon } from 'semantic-ui-react'

class GIFT extends Component {
    constructor(props){
    	super(props);
    	this.state = {person:'',present:''};
    }
    onChangePerson(e){
      this.setState({person:e.target.value})
    }
    onChangePresent(e){
      this.setState({present:e.target.value})
    }
    render() {
        return (
            <div className="GIFT">
              <Form>
                 <Form.Field>
                  <label>Person</label>
                  <Input
                  placeholder='person'
                  type='text'
                  value={this.state.person}
                  className='person-input'
                  onChange={this.onChangePerson.bind(this)}
                   />
                 </Form.Field>
                 <Form.Field>
                   <label>Present</label>
                    <Input
                    placeholder='present'
                    type='text'
                    className='present-input'
                    value={this.state.present}
                    onChange={this.onChangePresent.bind(this)}
                     />
                 </Form.Field>
                 <Icon
                 name='remove'
                 className='remove-btn'
                 onClick={() => this.props.removeClick(this.props.value.id)}
                 />
              </Form>
            </div>
        );
    }
}

export default GIFT
