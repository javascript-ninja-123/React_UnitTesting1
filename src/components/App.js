import React,{Component} from 'react';
//import from Semantic Ui -- front end UI
import { Button } from 'semantic-ui-react';
import Gift from './Gift';

class APP extends Component {
    constructor(props){
    	super(props);
    	this.state = {gifts:[]};
    }
    onClick(){
      let gifts = [...this.state.gifts];
      const id = gifts.map(value => value.id)
      const Max_id =id.length > 0 ? Math.max(...id) : 0;
      gifts.push({id:Max_id + 1});
      this.setState({gifts})
    }
    removeClick(id){
      const copyGift = [...this.state.gifts];
      let filteredGift = copyGift.filter(value => value.id !== id);
      this.setState({gifts:filteredGift})
    }
    renderDom(){
      return  this.state.gifts.map((value,index) => {
        return (
          <Gift key={index} removeClick={this.removeClick.bind(this)} value={value}/>
        )
      })
    }
    render() {
        return (
            <div className="APP">
                  <h2>Gift Giver</h2>
                  <div className='gift-list'>
                    {this.renderDom()}
                  </div>
                  <Button
                  className='add-button'
                  onClick={this.onClick.bind(this)}
                  >
                   Add
                 </Button>
            </div>
        );
    }
}

export default APP
