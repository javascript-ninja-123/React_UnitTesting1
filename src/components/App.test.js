import React from 'react';
import {shallow} from 'enzyme';
import setupTest from '../setupTest';
import App from './App';
import Gift from './Gift';

describe('App', () => {
  //define app
  const app = shallow(<App/>);
  const gift = shallow(<Gift />);

  //new describe for html
  describe('html (static)', () => {
    it('renders correctly', () => {
      expect(app).toMatchSnapshot();
    })
    it('starts with a state, an empty array list of gifts', () => {
      expect(app.state().gifts).toEqual([]);
    })
  })
  //ends here

  //new describe for javaScript
  describe('javascript (behavior)', () => {
    const id = 1;
    beforeEach(() => {
      app.find('.add-button').simulate('click');
    })
    afterEach(() => {
      app.setState({gifts:[]})
    })
    it('add a new gift when click a button, add-gift', () => {
      expect(app.state().gifts).toEqual([{id}])
    })
    it('add a new gift to the rendered list when Clicking the add gift', () => {
      expect(app.find('.gift-list').children()).toHaveLength(1);
    })
    it('create a  Gift component correctly', () => {
      expect(gift.exists()).toBe(true);
    })

    //remove a gift
    describe('User wants to remove the added gift', () => {
      beforeEach(() => {
        app.instance().removeClick(id);
      })
      it('removees the gift from state', () => {
        expect(app.state().gifts).toEqual([]);
      })
    })
    //ends here

  })
  //ends here

})
