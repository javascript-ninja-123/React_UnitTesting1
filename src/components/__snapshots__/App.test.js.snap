// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App html (static) renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <APP />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Gift Giver
</h2>,
        <div
          className="gift-list"
>
          
</div>,
        <Button
          as="button"
          className="add-button"
          onClick={[Function]}
>
          Add
</Button>,
      ],
      "className": "APP",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": "Gift Giver",
        },
        "ref": null,
        "rendered": "Gift Giver",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "gift-list",
        },
        "ref": null,
        "rendered": Array [],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "as": "button",
          "children": "Add",
          "className": "add-button",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Add",
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Gift Giver
</h2>,
          <div
            className="gift-list"
>
            
</div>,
          <Button
            as="button"
            className="add-button"
            onClick={[Function]}
>
            Add
</Button>,
        ],
        "className": "APP",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Gift Giver",
          },
          "ref": null,
          "rendered": "Gift Giver",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "gift-list",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {
            "as": "button",
            "children": "Add",
            "className": "add-button",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Add",
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
